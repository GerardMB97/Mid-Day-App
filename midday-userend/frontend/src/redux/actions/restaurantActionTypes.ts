const restaurantActionTypes = {
  LOAD_CATEGORIES: 'LOAD_CATEGORIES',
  FILTER_CATEGORIES: 'FILTER_CATEGORIES',
  LOAD_RESTAURANTS: 'LOAD_RESTAURANTS'
};

export default restaurantActionTypes
;
