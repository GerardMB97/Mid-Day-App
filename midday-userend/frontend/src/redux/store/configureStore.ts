import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import reduxImmutableStateInvariant from 'redux-immutable-state-invariant';
import rootReducer from '../reducers';
import initialState from './initialState';

function configureStore () {
  const composeEnhancers = compose;

  return createStore(
    rootReducer,
    initialState,
    composeEnhancers(applyMiddleware(reduxImmutableStateInvariant(), thunk))
  );
}

export default configureStore();
