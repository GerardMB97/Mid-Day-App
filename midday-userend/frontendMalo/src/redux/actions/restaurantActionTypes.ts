const restaurantActionTypes = {
  LOAD_CATEGORIES: 'LOAD_CATEGORIES'
};

export default restaurantActionTypes
;
