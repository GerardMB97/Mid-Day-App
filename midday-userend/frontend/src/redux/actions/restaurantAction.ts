import axios from 'axios';
import restaurantActionTypes from './restaurantActionTypes';

export const loadCategories = () => {
  return async (dispatch: Function) => {
    const categories = await axios.get('http://localhost:5000/api/midday/restaurants/categories');

    dispatch({
      type: restaurantActionTypes.LOAD_CATEGORIES,
      categories
    });
  };
}
;
