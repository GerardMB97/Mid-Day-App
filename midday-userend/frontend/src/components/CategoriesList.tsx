import React, { useEffect, useState } from 'react';
import { Dispatch, bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Icon from 'react-native-vector-icons/Ionicons';
import {
  StyleSheet,
  TextInput,
  Text,
  FlatList,
  View,
  SafeAreaView,
  ImageBackground
} from 'react-native';
import colors from '../../colors';
import { loadCategories, filterCategories } from '../redux/actions/restaurantAction';
import { Category } from '../models';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    marginTop: 100,
    alignItems: 'center',
    width: '100%',
    paddingLeft: 20,
    paddingRight: 20
  },
  input: {
    width: '100%',
    height: 60,
    backgroundColor: colors.green,
    justifyContent: 'center',
    paddingLeft: 60,
    marginBottom: 20,
    borderRadius: 10
  },
  title: {
    fontSize: 20,
    fontWeight: '700',
    marginBottom: 25
  },
  listContainer: {
    flex: 1,
    width: '100%'
  },
  list: {
    flexDirection: 'column'
  },
  listElement: {
    width: '48%',
    height: 150,
    borderRadius: 15,
    marginBottom: 15,
    borderColor: 'black',
    borderWidth: 1,
    overflow: 'hidden'
  },
  nameContainer: {
    backgroundColor: colors.transparentGray,
    width: '100%',
    height: 40,
    borderTopLeftRadius: 15,
    borderTopRightRadius: 15,
    justifyContent: 'center',
    alignItems: 'center'
  },
  search: {
    fontSize: 25,
    color: 'black',
    position: 'absolute',
    left: '8%',
    top: '2%'
  },
  image: {
    width: '100%',
    height: '100%',
    position: 'absolute',
    top: 0
  }
});
function CategoriesList ({ categories, actions }:any) {
  const [inputValue, setInputValue] = useState('');

  useEffect(() => {
    actions.loadCategories();
  }, []);
  return (
    <View style = {styles.container}>
      <TextInput
      style = {styles.input}
      placeholder="Tipo de menú o restaurante"
      value={inputValue}
      onChangeText={(text) => { setInputValue(text); actions.filterCategories(text); }}>
      </TextInput>
      <Icon style={styles.search} name= "search-outline"></Icon>
      <Text style = {styles.title}>¿Qué Menú te apetece hoy?</Text>
      <SafeAreaView style = {styles.listContainer}>
      <FlatList
        style = {styles.list}
        columnWrapperStyle={{ justifyContent: 'space-between' }}
        numColumns={2}
        data ={inputValue.length ? categories.filteredCategories : categories.allCategories}
        keyExtractor = {(item => item.name)}
        renderItem={({ item }) =>
          <View style = {styles.listElement} >
            <ImageBackground source= {{ uri: item.image }} style = {styles.image}>
              <View style = {styles.nameContainer}><Text>{item.name}</Text></View>
            </ImageBackground>
          </View>}
      >
      </FlatList>
      </SafeAreaView>
    </View>

  );
}

CategoriesList.propTypes = {
  categories: PropTypes.shape({
    allCategories: PropTypes.array.isRequired,
    filteredCategories: PropTypes.array.isRequired
  }).isRequired,
  actions: PropTypes.shape({
    loadCategories: PropTypes.func.isRequired,
    filterCategories: PropTypes.func.isRequired
  }).isRequired
};
interface destructuredSate{
  categories: Category[]
}
function mapStateToProps ({ categories }: destructuredSate) {
  return { categories };
}

function mapDispatchToProps (dispatch: Dispatch) {
  return { actions: bindActionCreators({ loadCategories, filterCategories }, dispatch) };
}

export default connect(mapStateToProps, mapDispatchToProps)(CategoriesList);
