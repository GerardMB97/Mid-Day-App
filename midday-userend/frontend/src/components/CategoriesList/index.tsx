import React, { useEffect } from 'react';
import { Dispatch, bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import {
  StyleSheet,
  Text,
  FlatList,
  View,
  SafeAreaView,
  ImageBackground,
  TouchableOpacity
} from 'react-native';
import colors from '../../../colors';
import { loadCategories, filterCategories } from '../../redux/actions/restaurantAction';
import { Category } from '../../models';

import SearchBar from '../SearchBar';
import NotFound from '../NotFound';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    marginTop: 100,
    alignItems: 'center',
    width: '100%',
    paddingLeft: 20,
    paddingRight: 20
  },
  input: {
    width: '100%',
    height: 60,
    backgroundColor: colors.green,
    justifyContent: 'center',
    paddingLeft: 60,
    marginBottom: 20,
    borderRadius: 10
  },
  title: {
    fontSize: 20,
    fontWeight: '700',
    marginBottom: 25
  },
  listContainer: {
    flex: 1,
    width: '100%'
  },
  list: {
    flexDirection: 'column'
  },
  listElement: {
    width: '48%',
    height: 150,
    borderRadius: 15,
    marginBottom: 15,
    borderColor: 'black',
    borderWidth: 1,
    overflow: 'hidden'
  },
  nameContainer: {
    backgroundColor: colors.transparentGray,
    width: '100%',
    height: 40,
    borderTopLeftRadius: 15,
    borderTopRightRadius: 15,
    justifyContent: 'center',
    alignItems: 'center'
  },
  search: {
    fontSize: 25,
    color: 'black',
    position: 'absolute',
    left: '8%',
    top: '2%'
  },
  image: {
    width: '100%',
    height: '100%',
    position: 'absolute',
    top: 0
  },
  nav: {
    width: '100%',
    height: 150
  }
});

function CategoriesList ({ categories, actions, navigation }:any) {
  const [inputValue, setInputValue] = React.useState('');

  useEffect(() => {
    if (!categories.allCategories.length) { actions.loadCategories(); }
  }, []);

  return (
    <View style = {styles.container}>
      <SearchBar inputValue={inputValue}
       setInputValue={setInputValue}
       inputPlaceholder='Tipo de menu o restaurante'
       action = {actions.filterCategories}
       ></SearchBar>
      <Text style = {styles.title}>¿Qué Menú te apetece hoy?</Text>
      <SafeAreaView style = {styles.listContainer}>
        {inputValue.length && !categories.filteredCategories.length
          ? <NotFound text="No hemos encontrado tu categoria"/>
          : <FlatList
        style = {styles.list}
        columnWrapperStyle={{ justifyContent: 'space-between' }}
        numColumns={2}
        data ={inputValue.length ? categories.filteredCategories : categories.allCategories}
        keyExtractor = {(item => item.name)}
        renderItem={({ item }) =>
          <View style = {styles.listElement} >

            <ImageBackground source= {{ uri: item.image }} style = {styles.image} >
              <TouchableOpacity testID={item.name} onPress={() => navigation.navigate('CategoriesDetail')} style = {styles.nav}>
                <View style={styles.nameContainer}>
                <Text>{item.name}</Text>
                </View>
                </TouchableOpacity>
            </ImageBackground>

          </View>}
      >
      </FlatList>
        }

      </SafeAreaView>
    </View>

  );
}

interface destructuredSate{
  categories: Category[]
}
function mapStateToProps ({ categories }: destructuredSate) {
  return { categories };
}

function mapDispatchToProps (dispatch: Dispatch) {
  return { actions: bindActionCreators({ loadCategories, filterCategories }, dispatch) };
}

export default connect(mapStateToProps, mapDispatchToProps)(CategoriesList);
