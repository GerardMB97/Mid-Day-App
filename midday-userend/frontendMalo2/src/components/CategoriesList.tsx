import React, { useEffect } from 'react';
import { Dispatch, bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import Icon from 'react-native-vector-icons/Ionicons';
import {
  StyleSheet,
  TextInput,
  Text,
  FlatList,
  View,
  SafeAreaView,
  ImageBackground,
  Image
} from 'react-native';
import colors from '../../colors';
import { loadCategories } from '../redux/actions/restaurantAction';
import { Category } from '../models';
function CategoriesList ({ categories, actions }:any) {
  useEffect(() => {
    actions.loadCategories();
  }, []);
  console.log(categories);

  return (
    <View style = {styles.container}>
      <Image source={require('../models/avatar.jpg')} style= {{ width: 200, height: 200 }}></Image>
      <TextInput style = {styles.input} placeholder="Tipo de menú o restaurante"></TextInput>
      <Icon style={styles.search} name= "search-outline"></Icon>
      <Text style = {styles.title}>¿Qué Menú te apetece hoy?</Text>
      <SafeAreaView style = {styles.listContainer}>
      <FlatList
        style = {styles.list}
        columnWrapperStyle={{ justifyContent: 'space-between' }}
        numColumns={2}
        data ={categories}
        keyExtractor = {(item => item.name)}
        renderItem={({ item }) =>
          <View style = {styles.listElement} >
            <ImageBackground source= {{ uri: item.image }} style = {styles.image}>
              <View style = {styles.nameContainer}><Text>{item.name}</Text></View>
            </ImageBackground>
          </View>}
      >
      </FlatList>
      </SafeAreaView>
    </View>

  );
}

CategoriesList.propTypes = {
  categories: PropTypes.array.isRequired,
  actions: PropTypes.shape({
    loadCategories: PropTypes.func.isRequired
  }).isRequired
};
interface destructuredSate{
  categories: Category[]
}
function mapStateToProps ({ categories }: destructuredSate) {
  return { categories };
}

function mapDispatchToProps (dispatch: Dispatch) {
  return { actions: bindActionCreators({ loadCategories }, dispatch) };
}

export default connect(mapStateToProps, mapDispatchToProps)(CategoriesList);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    marginTop: 100,
    alignItems: 'center',
    width: '100%',
    paddingLeft: 20,
    paddingRight: 20
  },
  input: {
    width: '100%',
    height: 60,
    backgroundColor: colors.green,
    justifyContent: 'center',
    paddingLeft: 60,
    marginBottom: 20,
    borderRadius: 10
  },
  title: {
    fontSize: 20,
    fontWeight: '700',
    marginBottom: 25
  },
  listContainer: {
    flex: 1,
    width: '100%'
  },
  list: {
    flexDirection: 'column'
  },
  listElement: {
    width: '48%',
    height: 150,
    borderRadius: 15,
    marginBottom: 15,
    borderColor: 'black',
    borderWidth: 1,
    overflow: 'hidden'
  },
  nameContainer: {
    backgroundColor: colors.transparentGray,
    width: '100%',
    height: 40,
    borderTopLeftRadius: 15,
    borderTopRightRadius: 15,
    justifyContent: 'center',
    alignItems: 'center'
  },
  search: {
    fontSize: 25,
    color: 'black',
    position: 'absolute',
    left: '8%',
    top: '2%'
  },
  image: {
    width: '100%',
    height: '100%',
    position: 'absolute',
    top: 0
  }
});
