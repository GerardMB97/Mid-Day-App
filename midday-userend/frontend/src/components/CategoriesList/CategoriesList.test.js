import React from 'react';
import { Provider } from 'react-redux';
import { fireEvent, render } from 'react-native-testing-library';

import CategoriesList from './index';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import * as actions from '../../redux/actions/restaurantAction';

jest.mock('../../redux/actions/restaurantAction');
jest.mock('react-native-vector-icons/Ionicons', () => 'Icon');

describe('Given a CategoriesList component', () => {
  const restaurants = {
    allRestaurants: [],
    categoryRestaurants: [],
    filteredRestaurants: []
  };
  const mockStore = configureStore([thunk]);
  const store = mockStore({ categories: { allCategories: [{ name: 'asian' }], filteredCategories: [] }, restaurants });

  beforeEach(() => {
    jest.spyOn(actions, 'filterCategories').mockReturnValueOnce({ type: '' });
    jest.spyOn(actions, 'loadCategories').mockReturnValueOnce({ type: '' });
    jest.spyOn(actions, 'loadRestaurants').mockReturnValueOnce({ type: '' });
  });
  test('Renders correctly', () => {
    const rendered = render(
      <Provider store={store}><CategoriesList /></Provider>
    );
    expect(rendered).toMatchSnapshot();
  });
  describe('When inputValue has length ', () => {
    test('Then should render the filtered categories', () => {
      const rendered = render(
      <Provider store={store}><CategoriesList /></Provider>
      );

      expect(rendered).toMatchSnapshot();
    });
  });
  describe('When a category is clicked', () => {
    test('Then it should navigate', () => {
      const navigation = {
        navigate: jest.fn()
      };

      const { getByTestId } = render(<Provider store={store}><CategoriesList navigation = {navigation}/></Provider>);
      fireEvent.press(getByTestId('asian'));

      expect(navigation.navigate).toHaveBeenCalled();
    });
  });
  describe('When inputValue has length', () => {
    test('Then it should render filteredCategories', () => {
      const realUseState = React.useState;
      const stubInitialState = 'asiatica';
      const store = mockStore({ categories: { allCategories: [], filteredCategories: [{ name: 'asian' }] }, restaurants });

      jest
        .spyOn(React, 'useState')
        .mockImplementationOnce(() => realUseState(stubInitialState));

      const rendered = render(<Provider store={store}><CategoriesList/></Provider>);
      expect(rendered).toMatchSnapshot();
    });
  });
  describe('When inputValue has length and filtered categories doesnt', () => {
    test('Then it should render Notfound component', () => {
      const realUseState = React.useState;
      const stubInitialState = 'asiatica';
      const store = mockStore({ categories: { allCategories: [], filteredCategories: [] }, restaurants: { allRestaurants: [{ name: 'elpepe' }], filteredRestaurants: [], categoryRestaurants: [] } });

      jest
        .spyOn(React, 'useState')
        .mockImplementationOnce(() => realUseState(stubInitialState));

      const rendered = render(<Provider store={store}><CategoriesList/></Provider>);
      expect(rendered).toMatchSnapshot();
    });
  });
  describe('When allCategories has no length', () => {
    test('Then loadCategories should be invoked', () => {
      const store = mockStore({ categories: { allCategories: [], filteredCategories: [] }, restaurants });

      render(<Provider store={store}><CategoriesList/></Provider>);
      expect(actions.loadCategories).toHaveBeenCalled();
    });
  });
});
