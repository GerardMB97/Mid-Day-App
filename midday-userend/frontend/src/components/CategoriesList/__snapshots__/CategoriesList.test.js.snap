// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a CategoriesList component Renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "paddingLeft": 20,
      "paddingRight": 20,
      "paddingTop": 100,
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "paddingLeft": 20,
        "paddingRight": 20,
        "width": "100%",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Tipo de menu o restaurante"
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "rgb(135, 175, 147)",
          "borderRadius": 10,
          "height": 60,
          "justifyContent": "center",
          "marginBottom": 20,
          "paddingLeft": 40,
          "width": "100%",
        }
      }
      testID="searchbar"
      underlineColorAndroid="transparent"
      value=""
    />
    <Icon
      name="search-outline"
      style={
        Object {
          "color": "black",
          "fontSize": 25,
          "left": "10%",
          "position": "absolute",
          "top": "20%",
        }
      }
    />
  </View>
  <Text
    style={
      Object {
        "fontSize": 20,
        "fontWeight": "700",
        "marginBottom": 25,
      }
    }
  >
    ¿Qué Menú te apetece hoy?
  </Text>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "flex": 1,
        "width": "100%",
      }
    }
  >
    <RCTScrollView
      data={
        Array [
          Object {
            "name": "asian",
          },
        ]
      }
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      style={
        Object {
          "borderRadius": 15,
          "flexDirection": "column",
        }
      }
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Array [
                Object {
                  "flexDirection": "row",
                },
                Object {
                  "justifyContent": "space-between",
                },
              ]
            }
          >
            <View
              style={
                Object {
                  "borderColor": "black",
                  "borderRadius": 15,
                  "borderWidth": 1,
                  "height": 150,
                  "marginBottom": 15,
                  "overflow": "hidden",
                  "width": "48%",
                }
              }
            >
              <View
                accessibilityIgnoresInvertColors={true}
                style={
                  Object {
                    "height": "100%",
                    "position": "absolute",
                    "top": 0,
                    "width": "100%",
                  }
                }
              >
                <Image
                  source={
                    Object {
                      "uri": undefined,
                    }
                  }
                  style={
                    Array [
                      Object {
                        "bottom": 0,
                        "left": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": 0,
                      },
                      Object {
                        "height": "100%",
                        "width": "100%",
                      },
                      undefined,
                    ]
                  }
                />
                <View
                  accessible={true}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "height": 150,
                      "opacity": 1,
                      "width": "100%",
                    }
                  }
                  testID="asian"
                >
                  <View
                    style={
                      Object {
                        "alignItems": "center",
                        "backgroundColor": "rgba(242, 242, 242, 0.8)",
                        "borderTopLeftRadius": 15,
                        "borderTopRightRadius": 15,
                        "height": 40,
                        "justifyContent": "center",
                        "width": "100%",
                      }
                    }
                  >
                    <Text>
                      asian
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </RCTSafeAreaView>
</View>
`;
